app/src/main/java/com/vestrel00/daggerbutterknifemvp/ui/common/BaseActivity.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2017 Vandolf Estrellado
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.vestrel00.daggerbutterknifemvp.ui.common;
+
+import android.app.Activity;
+import android.app.Fragment;
+import android.app.FragmentManager;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import dagger.android.AndroidInjection;
+import dagger.android.AndroidInjector;
+import dagger.android.DispatchingAndroidInjector;
+import dagger.android.HasFragmentInjector;
+
+/**
+ * Abstract Activity for all Activities to extend.
+ * <p>
+ * <b>DEPENDENCY INJECTION</b>
+ * We could extend {@link dagger.android.DaggerActivity} so we can get the boilerplate
+ * dagger code for free. However, we want to avoid inheritance (if possible and it is in this case)
+ * so that we have to option to inherit from something else later on if needed.
+ */
+public abstract class BaseActivity extends Activity implements HasFragmentInjector {
+
+    @Inject
+    @Named(BaseActivityModule.ACTIVITY_FRAGMENT_MANAGER)
```

Comment:
## Why use `@Named` vs `@Qualifier`?

Short answer is preference and lack of better naming ideas. 

Instead of using `@Named`, we can also use a custom [`Qualifier`](https://google.github.io/dagger/users-guide.html#qualifiers) in order to distinguish between the `FragmentManager` type from the activity versus the fragment.

E.G.

```java
@Qualifier
@Retention(RetentionPolicy.RUNTIME)
public @interface ActivityFragmentManager {
}

@Qualifier
@Retention(RetentionPolicy.RUNTIME)
public @interface FragmentFragmentManager {
}

@Module
public abstract class BaseActivityModule {

    @Provides
    @ActivityFragmentManager
    @PerActivity
    static FragmentManager activityFragmentManager(Activity activity) {
        return activity.getFragmentManager();
     }
}

@Module
public abstract class BaseFragmentModule {

    @Provides
    @FragmentFragmentManager
    @PerFragment
    static FragmentManager childFragmentManager(@Named(FRAGMENT) Fragment fragment) {
        return fragment.getChildFragmentManager();
    }
}

public abstract class BaseActivity extends Activity implements HasFragmentInjector {

    @Inject
    @ActivityFragmentManager
    protected FragmentManager fragmentManager;
    ...
}

public abstract class BaseFragment extends Fragment implements HasFragmentInjector {

    @Inject
    @FragmentFragmentManager
    protected FragmentManager fragmentManager;
    ...
}
```

I'm not a fan of the name `FragmentFragmentManager`. It looks weird (though doesn't sound as weird). I can't think of a more appropriate name... Besides, using `@Named` is common practice. If `@Qualifiers` were used every time multiple instances of the same type had to be provided in the same graph, then we would end up with large quantities of qualifiers (not that its bad or anything). It certainly has its advantages over using `@Named`, which on a side note is just another custom qualifier;

```java
@Qualifier
@Documented
@Retention(RUNTIME)
public @interface Named {
    String value() default "";
}
```

## Why use `@Qualifier` over `@Named`?

There are 2 concrete advantages of using `@Qualifier` over `@Named`, though be it a bit "elitist".

1. Solid `@interface` references instead of possibly brittle Strings.
2. `@Named` Strings may "leak" a reference to the module / component via an import.

We could easily mitigate these `@Named` issues though;

1. Using `@Named` requires the use of Strings, which can be misspelled. So a module may provide a "peach" but a variable may be injected with a "peech". This will never happen when using `@Qualifier`. However, a simple way to fix this flaw is to simply define the String (i.e. "peach") as a static final field and use that instead of typing the string over and over.

2. We mostly come to the immediate conclusion that we should place the static final String in the `@Module` that uses/provides it. At a glance, there seems to be no issue with this approach. However, if we really think hard and adopt a super strict policy, we'll notice that this will lead to a violation of core principle of dependency injection.

    That principle is that a class should not know anything about how it is injected or anything related to the injection process. Take the setup shown in this repository,

    ```java
    import com.vestrel00.daggerbutterknifemvp.ui.common.view.BaseFragmentModule;

    @PerFragment
    public final class PerFragmentUtil {

        @Inject
        PerFragmentUtil(@Named(BaseFragmentModule.FRAGMENT) Fragment fragment) {
        }
    }
    ```

    > Note that [`PerFragmentUitl` appears in the next PR](https://github.com/vestrel00/android-dagger-butterknife-mvp/pull/21/files#diff-63e5452481b67449a3e187ba6e5540b3).

    Notice the import for `com.vestrel00.daggerbutterknifemvp.ui.common.view.BaseFragmentModule;`, which is required to access the `FRAGMENT` String used for `@Named`. In turn, `PerFragmentUtil` now references the `BaseFragmentModule`, which it shouldn't. Classes being injected should have no visibility of `@Module`s and `@Component`s. This is clearly stated here: https://google.github.io/dagger/android.html#daggerandroid. It is one of the main motivations of why the *dagger.android* extension was created.

    ![Official Dagger & Android snippet](https://user-images.githubusercontent.com/4332753/33332569-3918ac10-d42a-11e7-8600-7e66312488dc.png)

    In practice, this violation is bad because not `PerFragmentUtil` can only be created / provided when `BaseFragmentModule` is a part of the injection graph. What if we want to use a different module other than `BaseFragmentModule`? We can't. Well we can as long as the String values are the same but still doesn't look right when we look at the code. We won't really encounter any issues in this particular example (`BaseActivityModule`, after all, is a module that is included in all of our modules so it is guaranteed to be in the injection graph), this was just an illustration of why this practice is bad.

    A simple solution is to move the static final String `FRAGMENT` outside of the `BaseActivityModule` into a separate class. This way, the `BaseActivityModule` need not be referenced. Then any `@Module` can be used in the injection graph to provide the `@Named(FRAGMENT`). The caveat to this solution is that a separate class would be required. Furthermore, the static final String seems like it should belong to `BaseActivityModule`.

This has been a long "elitist" rant. I recognize that these are really nit-picky details and won't really pose any issues during development. Therefore, we ignore these "issues" in this repo as they are far from pragmatic. We want to be pragmatic programmers, not elitists / over-engineerists!